nni/compression/pytorch/speedup/compressor.py

```diff
@@ -1,143 +1,336 @@
-                output_cmask = infer_from_inshape[m_type](module_masks,
-                                                          in_shape,
-                                                          self.torch_graph.name_to_node[module_name].auxiliary,
-                                                          last_module)
-            else:
-                output_cmask = infer_from_inshape[m_type](module_masks, in_shape)
-        if out_shape is not None:
-            _logger.debug("out_shape is not None")
-            if not m_type in infer_from_outshape:
-                raise RuntimeError(
-                    "Has not supported infering input shape from output shape for module/function: `{}`, {}"
-                    .format(m_type, module_name))
-            if m_type in ['aten::view', 'aten::flatten', 'aten::mean', 'aten::reshape']:
-                input_cmask = infer_from_outshape[m_type](module_masks, out_shape, self.torch_graph.name_to_node[module_name].auxiliary)
-            else:
-                input_cmask = infer_from_outshape[m_type](module_masks, out_shape)
+            weight_mask = None
+            if module_name in self.masks:
+                weight_mask = self.masks[module_name]
+            _, module = get_module_by_name(self.bound_model, module_name)
+            _auto_infer = AutoMaskInferenceClass(
+                module, dummy_input, in_masks, weight_mask, in_constants=in_constants,
+                state_dict=copy.deepcopy(module.state_dict()), batch_dim=self.batch_dim)
+        self.auto_inferences[unique_name] = _auto_infer
+        _auto_infer.name = node.unique_name
+        _auto_infer.fold_bias = self.fold_bias
 
-        if input_cmask:
-            predecessors = self.torch_graph.find_predecessors(module_name)
-            for _module_name in predecessors:
-                self.infer_module_mask(_module_name, module_name, out_shape=input_cmask)
-        if output_cmask:
-            successors = self.torch_graph.find_successors(module_name)
-            for _module_name in successors:
-                self.infer_module_mask(_module_name, module_name, in_shape=output_cmask)
+        _auto_infer.update_direct_sparsity()
+        # also save the input debug names into the auto_infer
+        _auto_infer.input_debugname = input_debugname
+        # update the mask tensor and the internal output of the submodules
+        # after manually unpack the tuple/list of tensors, the number of the outputs
+        # of each node should always be one(Except for the TupleUnpack node at the end
+        # of the whole model)
+        assert len(node.outputs) == 1, 'The number of the output should be one after the Tuple unpacked manually'
+
+        out_debugname = node.outputs[0]
+        # update the output mask into self.masks
+        self.masks[out_debugname] = _auto_infer.output_mask
+        self.constant[out_debugname] = _auto_infer.out_constant
```

Comment:
what is the meaning of `out_constant`?

Reply:
`out_constant` is the constant values of the corresponding output mask which is used as the input constants for the following layers.